### PR TITLE
Support for warmstarting

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py [hunk 1]
```diff
@@ -449,6 +454,36 @@ class GPFIFOSearcher(ModelBasedSearcher):
...
+    transfer_learning_model : str (optional)
+        See `transfer_learning_task_attr`. Specifies the surrogate model to be
+        used for transfer lerning:
+        - 'matern52_product': Kernel is product of Matern 5/2 (not ARD) on
+            `transfer_learning_task_attr` and Matern 5/2 (ARD) on the rest.
+            Assumes that data from same task are more closely related than
+            data from different tasks
+        - 'matern52_same': Kernel is Matern 5/2 (ARD) on the rest of the
+            variables, `transfer_learning_task_attr` is ignored. Assumes
+            that data from all tasks can be merged together
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:485]: What happens if you pass a large number of observations? Perhaps not for this PR but could we have a simple baseline to trim down to a reasonable number (for instance by sampling or grabing top ones

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py [hunk 2]
```diff
@@ -107,7 +108,8 @@ def _create_internal(
             num_initial_random_choices: int = DEFAULT_NUM_INITIAL_RANDOM_EVALUATIONS,
             initial_scoring: Optional[str] = None,
             cost_attr: Optional[str] = None,
-            resource_attr: Optional[str] = None):
+            resource_attr: Optional[str] = None,
+            filter_observed_data: Optional[ConfigurationFilter] = None):
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:112]: Can this one be added to the docstring? (I could not find anywhere)

File: syne_tune/optimizer/schedulers/searchers/bayesopt/datatypes/common.py [hunk 3]
```diff
@@ -30,6 +30,9 @@ def dictionarize_objective(x):
 Configuration = Dict[str, Hyperparameter]
 
 
+ConfigurationFilter = Callable[[Configuration], bool]
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/bayesopt/datatypes/common.py:33]: Would be useful to add 1-2 sentence comment on the context and how its used for warm-starting.

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py [hunk 4]
```diff
@@ -449,6 +454,36 @@ class GPFIFOSearcher(ModelBasedSearcher):
...
+        Used to support transfer HPO, where the state contains observed data
+        from several tasks, one of which is the active one. To this end,
+        `configspace` must contain a categorical parameter of name
+        `transfer_learning_task_attr`, whose range are all task IDs. Also,
+        `transfer_learning_active_task` must denote the active task, and
+        `transfer_learning_active_config_space` is used as
+        `active_config_space` argument in :class:`HyperparameterRanges` (this
+        allows us to use a narrower search space for the active task than for
+        the union of all tasks (`configspace` must be that).
+    transfer_learning_active_task : str (optional)
```
Comment (geoalgo) [inline syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:467]:
```
        allows us to use a narrower search space for the active task than for
        the union of all tasks (`configspace` must be that).
```
Can we say in one sentence on how this is done?

**Review by geoalgo — DISMISSED** [review]
LGTM, I left a few remarks on some comments that could improve readability.

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py (hunk 1 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:485]: True, this particular way of doing warm-starting does not scale well.

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py (hunk 2 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:112]: I added something to transfer_learning_task_attr docstring. The thing is this filter_observed_data is not something you set directly.

File: syne_tune/optimizer/schedulers/searchers/bayesopt/datatypes/common.py (hunk 3 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bayesopt/datatypes/common.py:33]: OK

File: syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py (hunk 4 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py:467]: I've added something. The reationale here is that the ranges in active_config_space may be smaller than what is needed to cover all the configs over all tasks. For example, say you want a HP to lie in a certain range, but you have some data where the range was larger. This allows you to use that old data.

**Review by geoalgo — APPROVED** [review]

**Pull request merged**